### PR TITLE
Use separate syntax kinds for *or patterns* and *and patterns*.

### LINKED ISSUE #42016: Please consider separate kinds for pattern combinators
Currently it's not convenient to distinguish pattern combinators. 

Following expressions, I suggest:

```diff
  <Node Name="BinaryPatternSyntax" Base="PatternSyntax">
-    <Kind Name="BinaryPattern"/>
+    <Kind Name="DisjunctivePattern"/>
+    <Kind Name="ConjunctivePattern"/>
    ...
  </Node>
  <Node Name="UnaryPatternSyntax" Base="PatternSyntax">
-    <Kind Name="UnaryPattern"/>
+    <Kind Name="NegatedPattern"/>
    ...
  </Node>
```
/cc @gafter

### PR DESCRIPTION
**Pull request by gafter**
Fixes #42016

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs [hunk 1]
```diff
@@ -75,7 +75,7 @@ private PatternSyntax ParseDisjunctivePattern(Precedence precedence, bool afterI
             {
                 var orToken = ConvertToKeyword(this.EatToken());
                 var right = ParseConjunctivePattern(precedence, afterIs, whenIsKeyword);
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs]: Not related to this PR, but I noticed we're parsing and-patterns with right-associativity. I suspect that's not by-design.

**Comment by gafter** [comment]
@RikkiGibson @agocke Can you please review this tiny change to the pattern-matching implementation?

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml [hunk 2]
```diff
@@ -1939,7 +1939,8 @@
     </Field>
   </Node>
   <Node Name="BinaryPatternSyntax" Base="PatternSyntax">
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml]: You probably want to remove `==` and `!=` from relational patterns too (a few lines above this).
The parser still accept those, perhaps good for error recovery?

**Comment by gafter** [comment]
@RikkiGibson Could you look at this please?

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml (hunk 2 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml]: Yes, present for error recovery

---
In reply to: [386532742](https://github.com/dotnet/roslyn/pull/42023#discussion_r386532742) [](ancestors = 386532742)

**Comment by gafter** [comment]
@RikkiGibson Could you look at this please?

**Comment by gafter** [comment]
@RikkiGibson Could you look at this please?

**Comment by RikkiGibson** [comment]
Sorry for being slow to get to this!

**Pull request merged**